docs/src/formula.md

```diff
@@ -14,167 +14,184 @@ fields with possibly heterogeneous types.  One of the primary goals of
+(or response) variables `y` from the right-hand size (or predictor) variables on
+the right `1 + a + b + c + b&c`.
+
+The left-hand side has one term `y` which means that the resopnse variable is
+the column from the data named `:y`.  The response can be accessed with the
+analogous `response(f, df)` function.
+
+The right hand side is made up of a number of different **terms**, separated by
+`+`: `1 + a + b + c + b&c`.  Each term corresponds to one or more columns in the
+generated model matrix: 
+
+* The first term `1` generates a constant or "intercept" column full of `1.0`s.
+* The next two terms `a` and `b` correspond to columns from the data table
+  called `:a`, `:b`, which both hold numeric data (`Float64` and `Int`
+  respectively).  Such **continuous terms** are converted to `Float64` and
+  copied to the model matrix
+* The term `c` corresponds to the `:c` column in the table, which is _not_
+  numeric, so it has been [contrast coded](@ref Modeling-categorical-data):
+  there are three unique values or levels, and the default coding scheme
+  ([`DummyCoding`](@ref)) generates an indicator variable for each level after
+  the first (e.g., `df[:c] .== "b"` and `df[:c] .== "a"`).
+* The last term `b&c` is an **interaction term**, and generates model matrix
+  columns for each _pair_ of columns generated by the `b` and `c` terms.
+  Columns are combined with elementwise multiplication.  Since `b` generates
+  only a single column and `c` two, `b&c` generates two columns, equivalent to
+  `df[:b] .* (df[:c] .== "b")` and `df[:b] .* (df[:c] .== "c")`.
+
+Because we often want to include both "main effects" (`b` and `c`) and
+interactions (`b&c`) of multiple variables, the formula DSL uses the `*`
+operator to denote this:
+
+```@repl 1
+@formula(y ~ 1 + a + b*c)
 ```
 
-(We trigger parsing of the formula using the internal `Terms` type to show how
-the `Formula` expands).
-
-This applies to higher-order interactions, too: `a*b*c` expands to the main
-effects, all two-way interactions, and the three way interaction `a&b&c`:
+Also note that the interaction operators `&` and `*` are distributive with the
+term separator `+`:
 
-```jldoctest
-julia> Formula(StatsModels.Terms(@formula(y ~ 1 + a*b*c)))
-Formula: y ~ 1 + a + b + c + a & b + a & c + b & c + &(a, b, c)
+```@repl 1
+@formula(y ~ 1 + (a + b) & c)
 ```
 
-Both the `*` and the `&` operators act like multiplication, and are distributive
-over addition:
+## Non-DSL calls
 
-```jldoctest
-julia> Formula(StatsModels.Terms(@formula(y ~ 1 + (a+b) & c)))
-Formula: y ~ 1 + a & c + b & c
+Any calls to functions that are not built into the DSL (or part of an
+[extension](@ref Internals-and-extending-the-formula-DSL) provided by a modeling
+package) are treated like normal Julia code, and evaluated elementwise:
 
-julia> Formula(StatsModels.Terms(@formula(y ~ 1 + (a+b) * c)))
-Formula: y ~ 1 + a + b + c + a & c + b & c
+```@repl 1
+modelmatrix(@formula(y ~ 1 + a + log(1+a) + length(c)), df)
```

Comment:
What is `length(c)` translated to in the matrix? It's a bit confusing for me since it will give a single constant, but the formula only allows `+1`.

Reply:
I was being a little cheeky.  It will compute the length of each string in the `:c` column and insert those values in a column (which in this case of course is perfectly collinear with the intercept...)

Reply:
Ah, of course. Maybe take another example which is more typical of common uses?

Reply:
I've now expanded this example with a separate example of a function you might (perhaps) want to actually use for strings, and explicitly flagged it as such...